Inline Search: Surface search corrections

interlineae opened the pull request:
This PR surfaces search corrections to ensure feature parity with existing Instant Search Functionality in our upcoming Inline Search implementation, which leverages the ES v1.3 API.

Because Inline Search uses the default WordPress search form, I've opted for an approach that injects our new markup via JS, and which also adds some default sensible styles for the ~34 most popular Atomic themes. I find this ideal as (1) it should adopt sensible default styling, assuming the theme has sensible styles, and (2) it doesn't make any hard commits with respect to the search template--we inject into what's there rather than dictate what can be there.

Example:

![CleanShot 2025-04-29 at 16 32 34](https://github.com/user-attachments/assets/8c97e7af-c855-4c69-9756-c2f401292329)

## Proposed changes:

> [!NOTE] 
> - Styles can be overridden by the user with custom CSS.
> - There is a new filter `jetpack_search_title_selectors` that allows the user to specify additional CSS selectors if their theme does not use the three default selectors that the Twenty Ten through to Twenty Twenty Five themes use.

### Other information:

This has documentation implications for the future. The new filter, if it goes ahead, will need to be documented.

## Jetpack product discussion

N/A

## Does this pull request change what data or activity we track or use?

No.

## Testing instructions:

- Either set `add_filter( 'jetpack_search_replace_classic', '__return_true', 20 );` somewhere (e.g. using `Code Snippets` plugin) or add an early `return true` to `should_replace_classic_search()`, or append `inline-search` as a query string to your search result page.
- Make sure to run `jp watch packages/search` or equivalent
- Run a new search with a misspelled word, e.g. `wordl`. You should see a result like the screenshot above.
- Run a new search using a correctly spelled word. You should not see any search corrections.
- Attempt other misspellings. You should not see any corrections for words that don't return a result on your test site.
- Add an additional CSS selector using the new filter, e.g.

```
add_filter( 'jetpack_search_title_selectors', function( $selectors ) {
    $selectors[] = '.custom-search-title';
    return $selectors;
} );
```

- Check that the new selector has been added with `console.log( 'Jetpack Search Title Selectors:', selectors );` somewhere in [corrected-query.js](https://github.com/Automattic/jetpack/pull/42473/files#diff-df1ae1ab2dfd4e8f1516ad551244331c7f8d3967ccdf2e1a90deafd901ffe134).

- Test with a selection of any of the following themes to ensure sensible display:

```
astra
hello-elementor
tsubaki
Divi
oceanwp
storefront
kadence
pub/zoologist
Avada
hevor
blocksy
generatepress
neve
pub/maywood
pub/shawburn
pub/lodestar
pub/barnsbury
pub/morden
pub/karuna
pub/baskerville-2
pub/dara
pub/twentytwentytwo
pub/twentytwenty
pub/twentytwentythree
pub/twentyseventeen
pub/creatio-2
pub/twentysixteen
pub/exford
pub/radcliffe-2
pub/attar
pub/assembler
pub/hever
pub/twentytwentyfive
pub/twentytwentyfour
```

interlineae committed:
Initial search correction commit

- Add two new methods, init_corrected_query_hooks() and update_search_options_with_correction()
  - These are likely to be changed

- Adjust search-results.jsx to (1) search against the correct query while (2) display informative text
  - This will be adjusted as the current functionality is to establish a foundation while ensuring things work

- Adds new styles for the "Did you mean?" line

interlineae committed:
changelog

interlineae committed:
Update wording/order

- Add proper "Did you mean?" text
- Remove code from last week's testing
- Prep for turning the search suggestion into a link

interlineae committed:
Initial linking of corrected search result

interlineae committed:
- Change order of "No results" and "Did you mean" text
- Update styling
- Rather than offering a "Did you mean?" choice, automatically try search suggestion

interlineae committed:
Merge remote-tracking branch 'origin/trunk' into enhance/surface-search-correction

interlineae committed:
Restore Instant Search Changes

Restoring Instance Search changes back to trunk as our changes are going to apply to Inline Search only.

interlineae committed:
Initial Inline Search changes

- Hook into pre_get_posts() for later displaying the corrected_query text
- Add a new setup_corrected_query_hooks() method which is to be modified in a future commit.
- Add a new maybe_use_corrected_query() method
- Add a new prepend_corrected_query_to_first_result() method to ensure that the corrected_query, if it's used, displays in the "Search results for:..." heading
- Add get_corrected_query_html() for outputting markup
- Add JS for injecting our new markup. This will be moved to its own file in a future commit.

Functionally:

- If a spelling error is corrected, and there is a result for the corrected spelling, the search form will return with two headings: (H1) "Search results for... { corrected_query }" followed by (H2) "No results found for... { original_query }"

- If a spelling error is discovered but there is no result for the corrected spelling, the search form will return with one headaing: (H1) "No results found for... { original_query }"

interlineae committed:
Cleanup

interlineae committed:
Cleanup and Refactor

- Refactor output_corrected_query_script() method
- New get_title_selectors() method
- New enqueue_corrected_query_script() method
- Refactor output_corrected_query_html() method
- Moved JS to new ../inline-search/js file
- Add a new filter to allow users to specify their own search title selectors if their theme does not use
one of the three selectors specified. These were checked across themes Twenty Twenty Five going back to
Twenty Ten.

interlineae committed:
Let us try adding some tests

interlineae committed:
Test JS

interlineae committed:
Revise and update our tests

interlineae committed:
Revert tests for inclusion in a latter PR

interlineae committed:
Revise methods further and make proper use of Assets::class

interlineae committed:
Modernise our JS

gibrown commented on projects/packages/search/src/inline-search/class-inline-search.php:
```diff
@@ -421,4 +426,115 @@ public function get_search_result(
+		$original_query = isset( $_GET['s'] ) ? sanitize_text_field( wp_unslash( $_GET['s'] ) ) : null; // phpcs:ignore WordPress.Security.NonceVerification.Recommended
+
+		if ( ! empty( $this->search_result['corrected_query'] ) ) {
+			$message = ! empty( $this->search_result['results'] )
```
I think the translation here may have a problem because it is separated into different calls.

I think the string needs to be 'No results for <strong>%s</strong>' so that the location of the search term can be moved around.

Also slightly confused why we need two different strings here depending on whether there are any results. I thought if we don't have any results then we should just not indicate anything about trying to correct the search term.

interlineae replied:
Ah, the two strings were vestigial from an earlier idea and I glossed over them. Fixed with https://github.com/Automattic/jetpack/pull/42473/commits/3f2d9a8691547e5472a2d218d62cee9f33d3430c.

Will update the language as I look into the potential translation issue 🤔

interlineae replied:
Re: translation, let me know if this is better https://github.com/Automattic/jetpack/pull/42473/commits/bde3ebbad6b2914f7e9ae070b52a4cfa2731ea7f

gibrown replied:
Yep. That looks better.

gibrown reviewed (commented):
Nice work. Haven't tested, just had the one comment.

Maybe @jeherve could also recommend someone with more theme experience since I think that is the riskiest part here and to also review that new filter.

interlineae committed:
Simply corrected query html

interlineae committed:
Refactor get_corrected_query_html()

- Check $_GET['s'] with null coalescing instead of isset
- Favour an early return instead of the ! checks
- Change message language to 'No results for %s'
- Remove <strong> tags to avoid forcing bold

I also looked into an option like inserting <span> tags in case a user wanted to
style the $original_query word themselves, but I wasn't happy with the implications
of this for translators, and the word can always be styled with a bit of
fairly straightforward JS.

interlineae committed:
Merge remote-tracking branch 'origin/trunk' into enhance/surface-search-correction

jeherve commented:
> Maybe @jeherve could also recommend someone with more theme experience since I think that is the riskiest part here and to also review that new filter.

I think you could ask folks currently working on our a8c themes. I believe @MaggieCabrera and the folks in Ignite have been involved with our themes lately.

MaggieCabrera commented on projects/packages/search/src/inline-search/js/corrected-query.js:
```diff
@@ -0,0 +1,26 @@
+	notice.innerHTML = html;
+
+	notice.className = `${ titleElement.className } ${ notice.className }`;
+	notice.style.cssText = 'font-size: 0.9em; margin-top: 10px; padding-top: 0;';
```
Inline styles are generally not great to use. In core we try to reserve them for user selected styles, so they are not overriden in the cascade. If this happens to work badly on any given theme, the theme can do nothing to fix that. A class would be a better option. I would also prefer an em or rem unit for the padding, if we must have one. We could also be adding a paragraph instead? I don't know because I am having a lot of trouble getting this PR in my local environment.

interlineae replied:
@MaggieCabrera If you keep having trouble, send me a DM and I'll take a look with you to see what might be going on.

interlineae replied:
@MaggieCabrera 

> Inline styles are generally not great to use. In core we try to reserve them for user selected styles, so they are not overriden in the cascade. If this happens to work badly on any given theme, the theme can do nothing to fix that. A class would be a better option. I would also prefer an em or rem unit for the padding, if we must have one. We could also be adding a paragraph instead? I don't know because I am having a lot of trouble getting this PR in my local environment.

So, the issue I'm trying to solve is the kind of styling struggle that a theme like Twenty Twenty Four has, given the current approach. If I plainly inject the markup it inherits some styling as so:

![CleanShot 2025-04-02 at 17 01 33](https://github.com/user-attachments/assets/1ce9a2c4-6601-4ce0-a645-809fa25c6688)

In this case:

```
.is-layout-constrained > :where(:not(.alignleft):not(.alignright):not(.alignfull)) {
    max-width: var(--wp--style--global--content-size);
    margin-left: auto !important;
    margin-right: auto !important;
}
```

I've written up a new `Webpack` config for the `inline-search/` directory and that seems to work to load a new stylesheet with a class, and so on (I stole it from what's already there for `instant-search/`). It works, there are no inline styles, and it can handle copying styles over more elegantly than copying over the parent element class (`page-title`, `archive-title`, etc.)

But maybe injecting HTML isn't the best approach to begin with? Or maybe this is a case of including styles for the various Twenty Twenty * themes, or, maybe I'm trying to correct for something that isn't on me to correct, or... did you have any thoughts?

MaggieCabrera replied:
ok, the CSS you are seeing isn't coming from the theme, but from core, and you are likely to have the same problem on every single block theme out there. I don't know how hard would this be, but the ideal solution for me would be to edit how the block works in the site editor to include the extra text, so the user or themer could position it or style it the way they see fit. Injecting html is subject to not work in expected ways, and hard for users or themers to fix. So your solution would have to consider that if you can't allow full flexibility (the content being editable in the site editor) for the user to style, at least the theme should be able to override the defaults that you end up setting. Having said that, most block themes these days don't add any CSS at all, or barely any, and rely on the block styling to achieve their desired results. Have in mind that my perspective is from working on core blocks, I have very little experience with how Jetpack blocks are built.

interlineae replied:
Okay, let me 🤔 about this a bit more, as I think at this point we'd ideally support both Block and classic themes.

trakos reviewed (commented):
I've tested this and it worked great on a bunch of twenty* themes I've tried - nice work. It didn't quite work on another theme I've tried (neve), but that's expected since it can't handle all themes.

From the other discussion it sounds like there are broader considerations around how things behave in block themes?

When reading the PR I had one hacky idea - in `maybe_use_corrected_query` returning something like:

```
return "{$this->search_result['corrected_query']} – no results for {$query}";
```

(with proper i18n added, of course). But I guess that can't be styled and might interfere with custom filters on users' blogs.

Another option could be to just add a filter/function to fetch the corrected query and let theme developers decide how/if to display it and/or adding some custom block instead of showing something by default. Though that wouldn't work out of the box.

anaemnesis commented:
@trakos @MaggieCabrera

> Another option could be to just add a filter/function to fetch the corrected query and let theme developers decide how/if to display it and/or adding some custom block instead of showing something by default. Though that wouldn't work out of the box.

I've been taking a look into this option. Looking at [Twenty Twenty Five](https://github.com/WordPress/twentytwentyfive) for example, I'm thinking we could do something like this in a theme's `functions.php` (or, provide it as part of Jetpack?):

```
/**
 * Register the search correction block.
 */
function twentytwentyfive_register_search_blocks() {
    if ( class_exists( 'Automattic\Jetpack\Search\Inline_Search' ) ) {
        register_block_type( 'jetpack/search-correction', [
            'render_callback' => 'twentytwentyfive_render_search_correction',
            'attributes'      => [],
        ]);
    }
}
add_action( 'init', 'twentytwentyfive_register_search_blocks' );

/**
 * Render the search correction block.
 */
function twentytwentyfive_render_search_correction() {
    if ( class_exists( 'Automattic\Jetpack\Search\Inline_Search' ) ) {
        $search = Automattic\Jetpack\Search\Inline_Search::instance();
        $html = $search->get_corrected_query_html();
        
        if ( ! empty( $html ) ) {
            return sprintf(
                '<div class="wp-block-jetpack-search-correction">%s</div>',
                $html
            );
        }
    }
    return '';
}
```

Then in a given theme's `search.html` template, `<!-- wp:jetpack/search-correction /-->` can be added to display the query correction:

```
<main class="wp-block-group" style="margin-top:var(--wp--preset--spacing--60)">
	<!-- wp:query-title {"type":"search"} /-->
	<!-- wp:jetpack/search-correction /-->
	<!-- wp:pattern {"slug":"twentytwentyfive/hidden-search"} /-->
	<!-- wp:pattern {"slug":"twentytwentyfive/template-query-loop"} /-->
	<!-- wp:pattern {"slug":"twentytwentyfive/more-posts"} /-->
</main>
```

Any styling concerns are then fully on the theme to resolve and we separate the functionality we provide here from its display.

For Classic themes we could have a similar approach, where a theme author would have the option of calling a function like `jetpack_search_display_correction();` in the `search.php` template to display the search results.

I'm not sure of another approach that is as flexible without forcing a `search.php` or `search.html` layout and I don't want to do that.

MaggieCabrera commented:
> @trakos @MaggieCabrera
> 
> > Another option could be to just add a filter/function to fetch the corrected query and let theme developers decide how/if to display it and/or adding some custom block instead of showing something by default. Though that wouldn't work out of the box.
> 
> I've been taking a look into this option. Looking at [Twenty Twenty Five](https://github.com/WordPress/twentytwentyfive) for example, I'm thinking we could do something like this in a theme's `functions.php` (or, provide it as part of Jetpack?):
> 
> ```
> /**
>  * Register the search correction block.
>  */
> function twentytwentyfive_register_search_blocks() {
>     if ( class_exists( 'Automattic\Jetpack\Search\Inline_Search' ) ) {
>         register_block_type( 'jetpack/search-correction', [
>             'render_callback' => 'twentytwentyfive_render_search_correction',
>             'attributes'      => [],
>         ]);
>     }
> }
> add_action( 'init', 'twentytwentyfive_register_search_blocks' );
> 
> /**
>  * Render the search correction block.
>  */
> function twentytwentyfive_render_search_correction() {
>     if ( class_exists( 'Automattic\Jetpack\Search\Inline_Search' ) ) {
>         $search = Automattic\Jetpack\Search\Inline_Search::instance();
>         $html = $search->get_corrected_query_html();
>         
>         if ( ! empty( $html ) ) {
>             return sprintf(
>                 '<div class="wp-block-jetpack-search-correction">%s</div>',
>                 $html
>             );
>         }
>     }
>     return '';
> }
> ```
> 
> Then in a given theme's `search.html` template, `<!-- wp:jetpack/search-correction /-->` can be added to display the query correction:
> 
> ```
> <main class="wp-block-group" style="margin-top:var(--wp--preset--spacing--60)">
> 	<!-- wp:query-title {"type":"search"} /-->
> 	<!-- wp:jetpack/search-correction /-->
> 	<!-- wp:pattern {"slug":"twentytwentyfive/hidden-search"} /-->
> 	<!-- wp:pattern {"slug":"twentytwentyfive/template-query-loop"} /-->
> 	<!-- wp:pattern {"slug":"twentytwentyfive/more-posts"} /-->
> </main>
> ```
> 
> Any styling concerns are then fully on the theme to resolve and we separate the functionality we provide here from its display.
> 
> For Classic themes we could have a similar approach, where a theme author would have the option of calling a function like `jetpack_search_display_correction();` in the `search.php` template to display the search results.
> 
> I'm not sure of another approach that is as flexible without forcing a `search.php` or `search.html` layout and I don't want to do that.

This could be a way, if you are ok with themes not doing that at all. TT themes will definitely not do that because they don't include plugin specific code at all. 3rd party themes and themes in general will most likely not go out of their way to do something like this. I think we need to abstract a solution that is universal and doesn't require the theme to add extra code, and if possible, that is easily overrideable (which was my main concern with the original solution).

I wonder if there's a better way to inject the code after the query title block? I would also insert a paragraph block instead of a simple div, which would probably play better with the core editor styles. So it essentially it would look like this on TT5 without extra styling:

<img width="1273" alt="Screenshot 2025-04-04 at 17 36 57" src="https://github.com/user-attachments/assets/ef810626-99f9-41f1-a40f-55b24d5b5ea6" />

gibrown commented:
Reading through this I'm thinking additional opinions may also help. Maybe @marktmcgrath can suggest someone from Team51 since I know that better Jetpack Search integration with themes is something they have asked for and wanted for a while.

marktmcgrath commented:
> Reading through this I'm thinking additional opinions may also help. Maybe @marktmcgrath can suggest someone from Team51 since I know that better Jetpack Search integration with themes is something they have asked for and wanted for a while.

@tommusrhodus may have an opinion on this. Either directly or by pulling others in from the team.

tommusrhodus commented:
> I wonder if there's a better way to inject the code after the query title block?

I'm in agreement with @MaggieCabrera here and would recommend using [Block Hooks](https://developer.wordpress.org/news/2024/03/exploring-the-block-hooks-api-in-wordpress-6-5/) to achieve this. JS injected markup is very likely to fail and have unexpected results.

anaemnesis commented:
@MaggieCabrera @tommusrhodus This approach looks fantastic, but I have a question.

When adding a `paragraph` block, the `No results for %s` string is easily overwritten in the editor. It would also be great if a user were able to remove / re-add the block if they decided to hide the search correction info, and then changed their mind.

Does that smell more of a "Search correction" block proper, rather than a `paragraph` tag hooked in?

interlineae committed:
Merge remote-tracking branch 'origin/trunk' into enhance/surface-search-correction

gibrown commented:
I'd like to propose that rather than trying to have a universal solution out of the gate we make it work for the top themes on Atomic. How about all the themes with over 1k active sites:
```
astra
hello-elementor
tsubaki
pub/twentytwentyfour
Divi
pub/assembler
pub/hever
pub/twentytwentyfive
oceanwp
storefront
pub/dara
pub/twentytwentytwo
pub/twentytwenty
pub/twentytwentythree
pub/twentyseventeen
pub/creatio-2
kadence
pub/zoologist
Avada
hevor
pub/exford
pub/radcliffe-2
pub/attar
blocksy
pub/twentysixteen
neve
pub/maywood
pub/shawburn
pub/lodestar
pub/barnsbury
pub/morden
generatepress
pub/karuna
pub/baskerville-2
```

Many of these can probably use similar solutions, but they obviously can't all use the same solution. I think our immediate goal is to ship something that works for 80% of cases and just have a plan for expanding the support over time. I imagine what this will look like is some case statement looking at the active theme and then we apply the solution that works best for that theme. Solutions for some themes may not be ideal, but we can always iterate on them.

I think we should also punt (or leave to a front end team) working on a new block for this functionality.

MaggieCabrera commented:
> @gibrown Would this be like a temporary compatibility file for approximately 34 themes? I'm kind of wondering if for the sake of this PR it should be split between getting the backend methods in place (this PR), and then figuring out the Block work in a subsequent PR. I think we'd also have similar concerns with the Search Branding PR (that doesn't yet exist) if we want that branding to be user removable.
> 
> I'm thinking a ~34 theme compat file is onerous, and doesn't really work around the Block issue unless for the sake of this file we were back to the injection method.
> 
> Thoughts?

I agree. Patching the themes is not great, even if we were to patch on the themes and not on the plugin, I'd rather go with the original injection method than doing this, since third party themes will not do this.

gibrown commented:
Sorry, I probably didn't explain it well. What I'm thinking is that if we have say three different methods of injecting corrections into themes let's say: theme_inject_1(); theme_inject_2(); and theme_inject_3(). Then could we do something like this:

```
// this could be cleaner :)

if ( apply_filter( 'jetpack-search-theme-injection') ) {
  // call the appropriate theme injection method
} else {
  switch( $theme ) {
    case "theme_a":
    case "theme_b":
      theme_inject_1();
      break;
    case "theme_c":
    case "theme_d":
    case "theme_e":
      theme_inject_2();
      break;
    case "theme_f":
    case "theme_g":
      theme_inject_3();
      break;
    default:
       // no corrections
  }
}
```

This is kinda related to how we have handled post meta with all the different plugins that we want to integrate with. It is tough to support all post meta that will ever be created by any plugin/theme, so we just make it easy to add support over time. See the post meta list here:

https://github.com/Automattic/jetpack/blob/trunk/projects/packages/sync/src/modules/class-search.php#L83

So when someone asks about why their theme doesn't support it we can do a single line change to add support. And as a short term hack they can always just add some filter code that chooses the injection method that they want.

We may still want to ship that original injection method, but maybe it will only get used for a small number of themes. I guess this is sort of similar to theme compatibility files (I don't know much about them) though maybe a lot lighter weight.

I'm also fine if this particular idea doesn't make sense. My overall point is that if we can find a way to support 80% of the themes on Atomic, and have a path for easily adding support to other themes in the future, then that would be a good outcome.

interlineae committed:
New class for search correction surfacing

Split out search correction methods into their own class

interlineae committed:
Create Inline Search Webpack config

We need a webpack config for Inline Search to allow us to add styles
sustainably, i.e., not inline. This is for the process of our interim "per theme"
approach for surfacing search corrections on the X most popular WPcom themes.

interlineae committed:
Add styling for default themes where necesssary:

pub/twentyten
pub/twentyeleven
pub/twentytwelve
pub/twentyfifteen
pub/twentysixteen
pub/twentyseventeen
pub/twentynineteen
pub/twentytwenty
pub/twentytwentytwo
pub/twentytwentythree
pub/twentytwentyfour
pub/twentytwentyfive

interlineae committed:
More theme styling

Styles for

pub/attar
pub/assembler
pub/hever
pub/dara
pub/creatio-2
pub/zoologist
pub/maywood
pub/shawburn
pub/lodestar
pub/barnsbury
pub/morden
pub/karuna
pub/baskerville-2
pub/exford
pub/radcliffe-2

interlineae committed:
Premium theme styling

Premium themes already make good use of the existing implementation. What we needed for them instead
is the addition of a few other selectors to target.

interlineae committed:
Merge remote-tracking branch 'origin/trunk' into enhance/surface-search-correction

anaemnesis commented:
> I'm also fine if this particular idea doesn't make sense. My overall point is that if we can find a way to support 80% of the themes on Atomic, and have a path for easily adding support to other themes in the future, then that would be a good outcome.

I've gone ahead with an approach in the vein of this idea, but different methods, a switch implementation, etc., were not required.

After checking what would be necessary across the 34 themes, I've instead:

1. Created a new webpack config for Inline Search, which I'm expecting will be helpful in the future beyond this PR.
2. Added a new `corrected-query.scss` stylesheet to hold styles for various themes that require adjustments for the injected HTML to display nicely.
3. Added a few new selectors for Premium themes

Both the `Divi` and `Avada` themes don't seem to be available on WPcom anymore so I'm figuring out how to test those before I re-open for review (but we can discuss on today's call).

interlineae committed:
Add query string to activate

There are some themes that have proven difficult to test, so let's add a
query string to enable Inline Search on demand so that we can merge this PR
test further, and iterate.

interlineae commented on projects/packages/search/tools/webpack.inline.config.js:
```diff
@@ -0,0 +1,58 @@
+const path = require( 'path' );
```
@kangzj @gibrown Recommended I reach out to you as I've not worked much with Webpack and put together this [config based on what we have for Instant Search](https://github.com/Automattic/jetpack/blob/trunk/projects/packages/search/tools/webpack.instant.config.js). Was hoping you could take a look for anything that could be improved?

interlineae commented on projects/packages/search/src/inline-search/class-inline-search.php:
```diff
@@ -18,10 +18,31 @@ class Inline_Search extends Classic_Search {
 	public static function should_replace_classic_search(): bool {
+		// Check for the inline search query parameter
+		// @TODO: Remove this once we go live with the new inline search
+		if ( isset( $_GET[ self::$inline_search_param ] ) ) { // phpcs:ignore WordPress.Security.NonceVerification.Recommended
```
This is for testing purposes post merge of this PR. This will be removed in a future PR.

gibrown reviewed (dismissed):
I think this is ready to go (I haven't been able to retest it because my dev environment is messed up, but I don't think we should wait on that). It also won't be live until we turn on inline search so should be low risk and we'll get another crack at testing as we turn that on (hopefully in a day or two).

trakos commented on projects/packages/search/src/inline-search/class-inline-search-correction.php:
```diff
@@ -0,0 +1,195 @@
+	/**
+	 * Get the search result from the Inline_Search instance.
+	 *
+	 * @return array|null The search result or null if not available.
```
It can also be an error


```suggestion
	 * @return array|\WP_Error|null The search result or null if not available.
```

trakos commented on projects/packages/search/src/inline-search/class-inline-search-correction.php:
```diff
@@ -0,0 +1,195 @@
+	 */
+	public function maybe_use_corrected_query( $query ) {
+		$search_result = $this->get_search_result();
+		if ( ! empty( $search_result['corrected_query'] ) && ! empty( $search_result['results'] ) ) {
```
I had an error when testing, and this triggered a fatal error `Cannot use object of type WP_Error as array`. We can avoid it by adding `is_array( $search_result) &&`:


```suggestion
		if ( is_array( $search_result) && ! empty( $search_result['corrected_query'] ) && ! empty( $search_result['results'] ) ) {
```

trakos commented on projects/packages/search/src/inline-search/class-inline-search-correction.php:
```diff
@@ -0,0 +1,195 @@
+		$original_query = sanitize_text_field( wp_unslash( $_GET['s'] ?? '' ) ); // phpcs:ignore WordPress.Security.NonceVerification.Recommended -- This is a search query.
+		$search_result  = $this->get_search_result();
+
+		if ( empty( $search_result['corrected_query'] ) || empty( $search_result['results'] ) ) {
```
I had an error when testing, and this triggered a fatal error `Cannot use object of type WP_Error as array`. We can avoid it by adding`! is_array( $search_result) ||`.

```suggestion
		if ( ! is_array( $search_result) || empty( $search_result['corrected_query'] ) || empty( $search_result['results'] ) ) {
```

trakos commented on projects/packages/search/src/inline-search/class-inline-search.php:
```diff
@@ -18,10 +18,31 @@ class Inline_Search extends Classic_Search {
+	 *
+	 * @var string
+	 */
+	private static $inline_search_param = 'inline-search';
```
If it's just fot testing, then would it make sense to use a more unique param, or checking for some expected value? There might be a conflict with other plugins using the same param.

interlineae replied:
I've removed this as we'd also need to programmatically disable instant search, and that was turning into a head ache. As this isn't really necessary (we have other ways to activate inline search when we need to), it's now gone.

trakos replied:
I think it's still in the PR for now? Just wanted to check if I'm understanding that right.

trakos commented on projects/packages/search/src/inline-search/class-inline-search-correction.php:
```diff
@@ -0,0 +1,195 @@
+	 * @return string The HTML for the corrected query notice or empty string if none.
+	 */
+	private function get_corrected_query_html() {
+		$original_query = sanitize_text_field( wp_unslash( $_GET['s'] ?? '' ) ); // phpcs:ignore WordPress.Security.NonceVerification.Recommended -- This is a search query.
```
I'm not sure if we should be using the `$_GET` parameter directly here. There could be some filters changing params, or rewriting from different param name. Perhaps something lke:

```
global $wp_query;
$original_query = $wp_query->get( 's' );
```

would work better?

interlineae replied:
Good shout, yes, I think this does.

trakos reviewed (commented):
Nice work! I tested this on the 2025, 2024, 2016, Astra, and Neve themes, and it worked well across all of them.

During testing, I briefly encountered a search error due to an issue with my sandbox environment. I noticed that it triggered a fatal error, so I’ve added a couple of comments with suggestions on how to handle that more gracefully.

My review focused primarily on the backend code.

interlineae committed:
Update projects/packages/search/src/inline-search/class-inline-search-correction.php

Co-authored-by: Piotr Stankowski <6437642+trakos@users.noreply.github.com>

interlineae committed:
Update projects/packages/search/src/inline-search/class-inline-search-correction.php

Co-authored-by: Piotr Stankowski <6437642+trakos@users.noreply.github.com>

interlineae committed:
Update projects/packages/search/src/inline-search/class-inline-search-correction.php

Co-authored-by: Piotr Stankowski <6437642+trakos@users.noreply.github.com>

interlineae committed:
Tidying up

- Ensure proper spacing before closing paranthesis
- Swap out $_GET for $wp_query->get to avoid potential issues with param rewrites, etc.

interlineae committed:
Remove inline-search query param and check

trakos reviewed (approved):
Thank you for all the updates and for working through this!

interlineae committed:
Merge remote-tracking branch 'origin/trunk' into enhance/surface-search-correction
